## Importing Pandas and Numpy libraries

In [1]:
import pandas as pd
import numpy as np

## We will be using the dataset 'concrete_data.csv'

The dataset mainly contains

1. Cement

2. Blast Furnace Slag

3. Fly Ash

4. Water

5. Superplasticizer

6. Coarse Aggregate

7. Fine Aggregate

## Reading the dataset into pandas dataframe

In [4]:
conc_data = pd.read_csv('concrete_data.csv')

In [6]:
conc_data.head()

Cement  Blast Furnace Slag  Fly Ash  Water  Superplasticizer  \
0   540.0                 0.0      0.0  162.0               2.5   
1   540.0                 0.0      0.0  162.0               2.5   
2   332.5               142.5      0.0  228.0               0.0   
3   332.5               142.5      0.0  228.0               0.0   
4   198.6               132.4      0.0  192.0               0.0   

   Coarse Aggregate  Fine Aggregate  Age  Strength  
0            1040.0           676.0   28     79.99  
1            1055.0           676.0   28     61.89  
2             932.0           594.0  270     40.27  
3             932.0           594.0  365     41.05  
4             978.4           825.5  360     44.30

The first concrete sample contains 540 cubic meters of cement, 0 cubic meters of blast furnace slag, 0 cubic meters of fly ash, 162 cubic meters of water, 2.5 cubic meters of superplasticizer, 1040 cubic meters of coarse aggregate, and 676 cubic meters of fine aggregate. After 28 days, this concrete mix achieves a compressive strength of 79.99 MPa.

The available data points

In [8]:
conc_data.shape

(1030, 9)

Checking Missing values

In [9]:
conc_data.describe()

Cement  Blast Furnace Slag      Fly Ash        Water  \
count  1030.000000         1030.000000  1030.000000  1030.000000   
mean    281.167864           73.895825    54.188350   181.567282   
std     104.506364           86.279342    63.997004    21.354219   
min     102.000000            0.000000     0.000000   121.800000   
25%     192.375000            0.000000     0.000000   164.900000   
50%     272.900000           22.000000     0.000000   185.000000   
75%     350.000000          142.950000   118.300000   192.000000   
max     540.000000          359.400000   200.100000   247.000000   

       Superplasticizer  Coarse Aggregate  Fine Aggregate          Age  \
count       1030.000000       1030.000000     1030.000000  1030.000000   
mean           6.204660        972.918932      773.580485    45.662136   
std            5.973841         77.753954       80.175980    63.169912   
min            0.000000        801.000000      594.000000     1.000000   
25%            0.000000        932.000000      730.950000     7.000000   
50%            6.400000        968.000000      779.500000    28.000000   
75%           10.200000       1029.400000      824.000000    56.000000   
max           32.200000       1145.000000      992.600000   365.000000   

          Strength  
count  1030.000000  
mean     35.817961  
std      16.705742  
min       2.330000  
25%      23.710000  
50%      34.445000  
75%      46.135000  
max      82.600000

In [10]:
conc_data.isnull().sum()

Cement                0
Blast Furnace Slag    0
Fly Ash               0
Water                 0
Superplasticizer      0
Coarse Aggregate      0
Fine Aggregate        0
Age                   0
Strength              0
dtype: int64

## Splitting data into predictors and target

The target variable in this problem is the concrete sample strength. Therefore, our predictors will be all the other columns.

In [12]:
conc_data_columns = conc_data.columns
predictors = conc_data[conc_data_columns[conc_data_columns != 'Strength']] 
target = conc_data['Strength']

In [13]:
predictors.head()

Cement  Blast Furnace Slag  Fly Ash  Water  Superplasticizer  \
0   540.0                 0.0      0.0  162.0               2.5   
1   540.0                 0.0      0.0  162.0               2.5   
2   332.5               142.5      0.0  228.0               0.0   
3   332.5               142.5      0.0  228.0               0.0   
4   198.6               132.4      0.0  192.0               0.0   

   Coarse Aggregate  Fine Aggregate  Age  
0            1040.0           676.0   28  
1            1055.0           676.0   28  
2             932.0           594.0  270  
3             932.0           594.0  365  
4             978.4           825.5  360

In [15]:
target.head()

0    79.99
1    61.89
2    40.27
3    41.05
4    44.30
Name: Strength, dtype: float64

In [16]:
n_cols = predictors.shape[1] 
n_cols

8

# IMPORTING KERAS

In [18]:
pip install keras

Note: you may need to restart the kernel to use updated packages.



[notice] A new release of pip is available: 24.0 -> 24.2
[notice] To update, run: python.exe -m pip install --upgrade pip


In [20]:
pip install --upgrade pip

   ---------------------------------------- 0.0/1.8 MB ? eta -:--:--
   ---------------------------------------- 0.0/1.8 MB ? eta -:--:--
    --------------------------------------- 0.0/1.8 MB 653.6 kB/s eta 0:00:03
   - -------------------------------------- 0.1/1.8 MB 770.8 kB/s eta 0:00:03
   -- ------------------------------------- 0.1/1.8 MB 731.4 kB/s eta 0:00:03
   -- ------------------------------------- 0.1/1.8 MB 654.9 kB/s eta 0:00:03
   --- ------------------------------------ 0.2/1.8 MB 756.6 kB/s eta 0:00:03
   ---- ----------------------------------- 0.2/1.8 MB 737.3 kB/s eta 0:00:03
   ---- ----------------------------------- 0.2/1.8 MB 724.0 kB/s eta 0:00:03
   ----- ---------------------------------- 0.2/1.8 MB 754.9 kB/s eta 0:00:03
   ------ --------------------------------- 0.3/1.8 MB 737.3 kB/s eta 0:00:03
   ------ --------------------------------- 0.3/1.8 MB 728.0 kB/s eta 0:00:03
   ------- -------------------------------- 0.3/1.8 MB 722.1 kB/s eta 0:00:03
   -

In [23]:
!pip install tensorflow

In [27]:
import tensorflow as tf
from tensorflow import keras

11490434/11490434 ━━━━━━━━━━━━━━━━━━━━ 17s 1us/step


C:\Users\mathe\AppData\Local\Programs\Python\Python39\lib\site-packages\keras\src\layers\reshaping\flatten.py:37: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(**kwargs)


Epoch 1/5
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 10s 4ms/step - accuracy: 0.8588 - loss: 0.4904
Epoch 2/5
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 8s 4ms/step - accuracy: 0.9543 - loss: 0.1528
Epoch 3/5
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 10s 4ms/step - accuracy: 0.9666 - loss: 0.1112
Epoch 4/5
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 8s 4ms/step - accuracy: 0.9727 - loss: 0.0879
Epoch 5/5
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 8s 4ms/step - accuracy: 0.9766 - loss: 0.0714
313/313 ━━━━━━━━━━━━━━━━━━━━ 1s 3ms/step - accuracy: 0.9751 - loss: 0.0823


[0.06806258857250214, 0.9794999957084656]

In [38]:
import tensorflow as tf
from tensorflow.keras import backend as K

print(K.backend())  # Should output "tensorflow"

tensorflow


In [39]:
import keras

 importing rest of the packages from the Keras library that we will need to build our regressoin model.

In [31]:
from keras.models import Sequential
from keras.layers import Dense

In [32]:
def regression_model():
    # create model
    model = Sequential()
    model.add(Dense(10, activation='relu', input_shape=(n_cols,)))
    model.add(Dense(1))
    model.compile(optimizer='adam', loss='mean_squared_error')
    return model

The function above constructs a model with a single hidden layer containing 10 neurons and utilizes a ReLU activation function. The model employs the Adam optimizer and mean squared error as the loss function.

Next, let's import scikit-learn to randomly divide the data into training and test sets.

In [33]:
from sklearn.model_selection import train_test_split

In [34]:
X_train, X_test, y_train, y_test = train_test_split(predictors, target, test_size=0.3, random_state=42)

## Training and Testing the Network

In [35]:
model = regression_model()

C:\Users\mathe\AppData\Local\Programs\Python\Python39\lib\site-packages\keras\src\layers\core\dense.py:87: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


In [41]:
epochs = 50
model.fit(X_train, y_train, epochs=epochs, verbose=1)

Epoch 1/50
23/23 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - loss: 110.3635
Epoch 2/50
23/23 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - loss: 102.7021
Epoch 3/50
23/23 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - loss: 105.2368 
Epoch 4/50
23/23 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - loss: 105.9137
Epoch 5/50
23/23 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - loss: 101.7202
Epoch 6/50
23/23 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - loss: 107.4221
Epoch 7/50
23/23 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - loss: 113.3169
Epoch 8/50
23/23 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - loss: 111.3750
Epoch 9/50
23/23 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - loss: 109.2938
Epoch 10/50
23/23 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - loss: 98.9217
Epoch 11/50
23/23 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - loss: 118.1305 
Epoch 12/50
23/23 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - loss: 102.7189
Epoch 13/50
23/23 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - loss: 113.5650
Epoch 14/50
23/23 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - loss: 108.8830 
Epoch 15/50
23/23 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - lo

In [42]:
loss_val = model.evaluate(X_test, y_test)
y_pred = model.predict(X_test)
loss_val

10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - loss: 96.6077   
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 8ms/step


102.05416107177734

Computing the mean squared error between the predicted concrete strength and the actual concrete strength.

importing the mean_squared_error function

In [47]:
from sklearn.metrics import mean_squared_error

In [48]:
mean_square_error = mean_squared_error(y_test, y_pred)
mean = np.mean(mean_square_error)
standard_deviation = np.std(mean_square_error)
print(mean, standard_deviation)

102.05417106594335 0.0


Creating a list of 50 mean squared errors and report mean and the standard deviation of the mean squared errors.

In [49]:
total_mean_squared_errors = 50
epochs = 50
mean_squared_errors = []
for i in range(0, total_mean_squared_errors):
    X_train, X_test, y_train, y_test = train_test_split(predictors, target, test_size=0.3, random_state=i)
    model.fit(X_train, y_train, epochs=epochs, verbose=0)
    MSE = model.evaluate(X_test, y_test, verbose=0)
    print("MSE "+str(i+1)+": "+str(MSE))
    y_pred = model.predict(X_test)
    mean_square_error = mean_squared_error(y_test, y_pred)
    mean_squared_errors.append(mean_square_error)

mean_squared_errors = np.array(mean_squared_errors)
mean = np.mean(mean_squared_errors)
standard_deviation = np.std(mean_squared_errors)

print('\n')
print("Below is the mean and standard deviation of " +str(total_mean_squared_errors) + " mean squared errors without normalized data. Total number of epochs for each training is: " +str(epochs) + "\n")
print("Mean: "+str(mean))
print("Standard Deviation: "+str(standard_deviation))

MSE 1: 89.6259536743164
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step 
MSE 2: 103.94611358642578
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step 
MSE 3: 56.8680419921875
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step 
MSE 4: 55.403526306152344
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step 
MSE 5: 53.9409065246582
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step 
MSE 6: 53.772396087646484
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step 
MSE 7: 62.6894416809082
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step 
MSE 8: 48.50874710083008
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step 
MSE 9: 52.098609924316406
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step 
MSE 10: 50.87533950805664
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 7ms/step
MSE 11: 48.86411666870117
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step 
MSE 12: 51.172603607177734
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step 
MSE 13: 57.57801055908203
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step 
MSE 14: 53.854942321777344
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step 
MSE 15: 49.05501174926758
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step 
MSE 16: 41.7466506